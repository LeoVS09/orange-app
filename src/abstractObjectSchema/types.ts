
export enum AosFieldType {
    Any = 'Any',
    OneToOne = 'OneToOne',
    OneToMany = 'OneToMany',
    String = 'String',
    Number = 'Number',
    Boolean = 'Boolean',
    Date = 'Date'
}

/**
 * Note: Actually Any field type is just define unknown about type,
 * but in base case we will mean that is simple type
 * if not one set it as Relation type
 */

export type AosSimeplFieldType =
    AosFieldType.Any |
    AosFieldType.String |
    AosFieldType.Number |
    AosFieldType.Boolean |
    AosFieldType.Date

export type AosRelationsFieldType = AosFieldType.OneToOne | AosFieldType.OneToMany

export interface AosSimpleField {
    type: AosSimeplFieldType
}

export interface AosRelationsField {
    type: AosRelationsFieldType
    schema: AosSchema
}

export type AosField = AosSimpleField | AosRelationsField

/**
 * Recursive scheme of object and relations to other objects
 * Like AST, but for object relations
 * */
export interface AosSchema {
    [keys: string]: AosField
}

// Plain Scheme of entity in database
export interface AosEntitySchema {
    primaryKey: string
    foreignKeys: Array<string>
    fields: AosEntityFields
}

// Map of entity fields
export interface AosEntityFields {
    [key: string]: AosFieldType | AosEntityField
}

export interface AosEntityField {
    type: AosFieldType
    table?: string
}

// Describe database relations scheme
export interface AosEntitySchemaStorage {
    [entity: string]: AosEntitySchema
}
