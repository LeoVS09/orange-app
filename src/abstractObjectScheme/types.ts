
export enum AosFieldType {
    Any = 'Any',
    OneToOne = 'OneToOne',
    OneToMany = 'OneToMany',
    String = 'String',
    Number = 'Number',
    Boolean = 'Boolean',
    Date = 'Date'
}

/**
 * Note: Actually Any field type is just define unknown about type,
 * but in base case we will mean that is simple type
 * if not one set it as Relation type
 */

export interface AosSimpleField {
    type:
        AosFieldType.Any |
        AosFieldType.String |
        AosFieldType.Number |
        AosFieldType.Boolean |
        AosFieldType.Date
}

export interface AosRelationsField {
    type : AosFieldType.OneToOne | AosFieldType.OneToMany
    scheme: AosScheme
}

export type AosField = AosSimpleField | AosRelationsField

// Recursive scheme of object and relations to other objects
export interface AosScheme {
    [keys: string]: AosField
}

// Plain Scheme of entity in database
export interface AosEntitySchema {
    primaryKey: string
    foreignKeys: Array<string>
    fields: AosEntityFields
}

// Map of entity fields
export interface AosEntityFields {
    [key: string]: AosFieldType
}
