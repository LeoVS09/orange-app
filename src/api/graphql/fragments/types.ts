export interface ResponseDataTest {
   id: string
   index: number
   input: string
   output: string
   isPublic: boolean
   createdAt: Date
   updatedAt: Date
}

export interface ResponseDataPartialProfile {
   id: string
   isTeacher: boolean
   firstName: string
   lastName: string
   user: {
      id: string
      name: string
   }
}

export interface ResponseDataTag {
   id: string
   name: string
}

export interface ResponseDataPartialProblem {
   id: string
   name: string
   createdAt: Date
   updatedAt: Date
   publicationDate: Date
   difficulty: number
   author: ResponseDataPartialProfile
   tester: ResponseDataPartialProfile
   problemsTags: {
      nodes: Array<{
         tag: ResponseDataTag
      }>
   }
}

export interface ResponseDataInputOutputType {
   id: string
   name: string
   code: string
}

export interface ResponseDataProblem extends ResponseDataPartialProblem {
   description: string
   inputDescription: string
   outputDescription: string
   note: string

   inputType: ResponseDataInputOutputType
   outputType: ResponseDataInputOutputType

   limitTime: number
   limitMemory: number

   tests: {
      nodes: Array<ResponseDataTest>
   }
}
