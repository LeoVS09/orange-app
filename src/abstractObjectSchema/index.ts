export * from './types'
export * from './guards'

// TODO: may be use graph representation of object scheme
// TODO: may be use for database store graph representation
// TODO: try to implement this with JSOG: https://github.com/jsog/jsog
// and typescript implementation https://github.com/emundo/jsog-typescript
