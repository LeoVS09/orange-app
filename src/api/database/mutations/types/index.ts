export * from './createInputType';
export * from './createOutputType';
export * from './createProblem';
export * from './createTag';
export * from './createTest';
export * from './deleteInputType';
export * from './deleteOutputType';
export * from './deleteProblem';
export * from './deleteTag';
export * from './deleteTest';
export * from './Login';
export * from './Register';
export * from './updateInputType';
export * from './updateOutputType';
export * from './updateProblem';
export * from './updateTag';
export * from './updateTest';
export * from './createCity';
export * from './createCountry';
export * from './createUniversity';
export * from './deleteCity';
export * from './deleteCountry';
export * from './deleteUniversity';
export * from './updateCity';
export * from './updateCountry';
export * from './updateUniversity';
export * from './createContest';
export * from './updateContest';
export * from './deleteContest';
