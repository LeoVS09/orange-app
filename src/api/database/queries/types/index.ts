export * from './Cities';
export * from './City';
export * from './Countries';
export * from './Country';
export * from './CurrentUser';
export * from './InputOutputTypes';
export * from './Problem';
export * from './Problems';
export * from './SearchCountries';
export * from './Tags';
export * from './University';
export * from './Universities';
export * from './CodeEditor';
export * from './CodeEditors';
export * from './Compiler';
export * from './Compilers';
export * from './Contest';
export * from './Contests';
export * from './Profile';
export * from './Profiles';
export * from './ProgrammingLanguage';
export * from './ProgrammingLanguages';
export * from './Team';
export * from './Teams';
export * from './Tag';
export * from './Tests';
