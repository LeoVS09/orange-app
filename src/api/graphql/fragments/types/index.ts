export * from './FullCity'
export * from './FullCountry'
export * from './FullProblem'
export * from './FullProfile'
export * from './FullUser'
export * from './InputType'
export * from './OutputType'
export * from './PartialCity'
export * from './PartialCountry'
export * from './PartialProblem'
export * from './PartialProfile'
export * from './PartialUniversity'
export * from './PartialUser'
export * from './PartialTag'
export * from './FullTag'
export * from './Test'
