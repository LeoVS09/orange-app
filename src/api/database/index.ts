import * as globalTypes from './global-types'
import * as queriesTypes from './queries/types'
import * as mutationsTypes from './mutations/types'

export * from './mutations'
export * from './queries'

export const types = {
  globalTypes,
  queriesTypes,
  mutationsTypes,
}

// TODO: add pre-compiling queries when build project
