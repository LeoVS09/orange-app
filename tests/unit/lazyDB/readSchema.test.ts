import cons from '../utils/console.mock'
import { AosSchema, AosFieldType } from '@/abstractObjectSchema'
import { ModelEventGetPropertyPayload } from '@/lazyDB/core/types'
import { appendPropertyToSchema } from '@/lazyDB/database/readSchema'
import { IDatabaseModelProducerStore } from '@/lazyDB/database/types'

const simplePayload = (name: string, type = AosFieldType.Any): ModelEventGetPropertyPayload => ({
    name, type, store: {} as IDatabaseModelProducerStore
})

describe('AOS builder', () => {
    let schema: AosSchema

    beforeEach(() => {
        schema = {}
    })

    it('append simple string property', () => {
        const payload = simplePayload('simple')

        const isAppended = appendPropertyToSchema(schema, payload)

        expect({ isAppended, schema }).toMatchSnapshot()
    })

    it('append multiple string properties', () => {
        const first = simplePayload('first')
        const second = simplePayload('second')
        const third = simplePayload('third')

        let isAppended = appendPropertyToSchema(schema, first)
        expect({ isAppended, schema }).toMatchSnapshot()

        isAppended = appendPropertyToSchema(schema, second)
        expect({ isAppended, schema }).toMatchSnapshot()

        isAppended = appendPropertyToSchema(schema, third)
        expect({ isAppended, schema }).toMatchSnapshot()
    })
})