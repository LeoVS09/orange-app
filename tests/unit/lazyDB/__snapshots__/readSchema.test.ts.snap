// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AOS builder append multiple string properties 1`] = `
Object {
  "isAppended": true,
  "schema": Object {
    "first": Object {
      "type": "Any",
    },
    "id": Object {
      "type": "Any",
    },
    "nodeId": Object {
      "type": "Any",
    },
  },
}
`;

exports[`AOS builder append multiple string properties 2`] = `
Object {
  "isAppended": true,
  "schema": Object {
    "first": Object {
      "type": "Any",
    },
    "id": Object {
      "type": "Any",
    },
    "nodeId": Object {
      "type": "Any",
    },
    "second": Object {
      "type": "Any",
    },
  },
}
`;

exports[`AOS builder append multiple string properties 3`] = `
Object {
  "isAppended": true,
  "schema": Object {
    "first": Object {
      "type": "Any",
    },
    "id": Object {
      "type": "Any",
    },
    "nodeId": Object {
      "type": "Any",
    },
    "second": Object {
      "type": "Any",
    },
    "third": Object {
      "type": "Any",
    },
  },
}
`;

exports[`AOS builder append simple string property 1`] = `
Object {
  "isAppended": true,
  "schema": Object {
    "id": Object {
      "type": "Any",
    },
    "nodeId": Object {
      "type": "Any",
    },
    "simple": Object {
      "type": "Any",
    },
  },
}
`;
